llava/llava.py

```diff
@@ -0,0 +1,196 @@
+        )
+        # select the hidden states from the desired layer
+        selected_image_feature = hidden_states[self.vision_feature_layer]
+
+        if self.vision_feature_select_strategy == "default":
+            selected_image_feature = selected_image_feature[:, 1:]
+        elif self.vision_feature_select_strategy == "full":
+            selected_image_feature = selected_image_feature
+        else:
+            raise ValueError(
+                f"Unexpected select feature strategy: {self.vision_feature_select_strategy}"
+            )
+        # pass image features through the multi-modal projector
+        image_features = self.multi_modal_projector(selected_image_feature)
+        # insert special image tokens in the input_ids
+        final_inputs_embeds = self._merge_input_ids_with_image_features(
+            image_features, inputs_embeds, input_ids
+        )
+
+        return final_inputs_embeds
+
+    def _merge_input_ids_with_image_features(
+        self, image_features, inputs_embeds, input_ids
+    ):
+
+        image_features = np.array(image_features)
+        inputs_embeds = np.array(inputs_embeds)
+        input_ids = np.array(input_ids)
+
+        _, num_image_patches, embed_dim = image_features.shape
+        batch_size, sequence_length = input_ids.shape
+
+        special_image_token_mask = input_ids == self.config.image_token_index
+        num_special_image_tokens = np.sum(special_image_token_mask, axis=-1)
+
+        # if no special image tokens found, return a warning
+        if np.all(num_special_image_tokens == 0):
+            logging.warning(
+                "No special image tokens found in the input. Please make sure to include <image> in your prompt."
+            )
+
+        # calculate the final sequence length. Will be the original sequence length + the # of image tokens to be inserted in.
+        final_sequence_length = (
+            np.max(num_special_image_tokens) * (num_image_patches - 1)
+        ) + sequence_length
+
+        non_image_indices = np.where(
+            input_ids != self.config.image_token_index)
+
+        new_token_positions = (
+            np.cumsum((special_image_token_mask *
+                      (num_image_patches - 1) + 1), axis=-1)
+            - 1
+        )
+        text_to_overwrite = new_token_positions[non_image_indices]
+
+        final_embedding = np.zeros(
+            (batch_size, final_sequence_length,
+             embed_dim), dtype=inputs_embeds.dtype
+        )
+
+        final_embedding[non_image_indices[0], text_to_overwrite, :] = inputs_embeds[
+            non_image_indices
+        ]
+
+        image_to_overwrite = np.all(final_embedding == 0, axis=-1)
+        reshaped_image_features = image_features.reshape(-1, embed_dim)
+        final_embedding[image_to_overwrite, :] = reshaped_image_features[
+            : np.sum(image_to_overwrite)
+        ]
+
+        return mx.array(final_embedding)
```

Comment:
@awni, this is the part where we have to use NumPy as a workaround because MLX doesn't support boolean indexing. Maybe you could give us some pointers on how we can implement this using MLX.

Reply:
1, you are right. I skipped the error checking for mismatched image tokens and image batches to simplify the implementation, but we could definitely add it. FYI: https://github.com/huggingface/transformers/blob/main/src/transformers/models/llava/modeling_llava.py#L333-L337
2. My understanding is that as long as the total number of images matches the total image tokens in a batch, it should work with batches. Given the final embedding is batched, however, I have not done any batch testing.

Reply:
Regarding 2, makes sense! Does the rest of the code work for batched text? E.g. the generation part properly handles different text examples? If it's just using our MLX LM llama implementation (which it looks like) then the answer is probably no. I don't know if that simplifies much though.

Reply:
Correct me if I am wrong, but I thought the Llama example should be able to handle batched text generation. The only difference is that the initial prompt will get the embedding with text and image from the code above and directly go to LLM's encode layer without going through the embedding/norm layer. Once that forward pass is completed, we get the next token logits and key-value cache, then the rest of process is just a standard forward pass for LLM.

Reply:
```
    def _merge_input_ids_with_image_features(
        self, image_features, inputs_embeds, input_ids
    ):
        image_token_index = self.config.image_token_index
        num_images, num_image_patches, embed_dim = image_features.shape

        # Positions of <image> tokens in `input_ids` and assume batch size is 1
        image_positions = [
            idx
            for idx, token_id in enumerate(input_ids[0])
            if token_id == image_token_index
        ]

        if len(image_positions) != num_images:
            raise ValueError(
                f"The input provided to the model are wrong. The number of image tokens is {len(image_positions)} while"
                f" the number of image given to the model is {num_images}."
            )

        text_segments = []
        start_idx = 0

        # Split the input embeddings into text segments based on the positions of <image> tokens
        for position in image_positions:
            text_segments.append(inputs_embeds[:, start_idx:position])
            start_idx = position + 1

        # Add the remaining text embeddings after the last <image> token
        if start_idx < inputs_embeds.shape[1]:
            text_segments.append(inputs_embeds[:, start_idx:])

        # Reshape image feature from (num_images, num_image_patches, embed_dim) to (num_images*num_image_patches*, embed_dim)
        image_embeddings = image_features.reshape(-1, image_features.shape[-1])

        final_embeddings = []
        for i, text_segment in enumerate(text_segments):
            final_embeddings.append(text_segment[0])
            if i < len(image_positions):
                # Add a slice of image embeddings corresponding to the current position.
                # This effectively replaces one <image> token with its associated image embeddings.
                final_embeddings.append(
                    image_embeddings[i : i + num_image_patches + 1]
                ) 

        # This creates a final embeding in shape (1, num_image_patches + num_text_tokens, embed_dim) representing the merged sequence of text and image embeddings.  
        final_embeddings = mx.concatenate(final_embeddings, axis=0).reshape(
            1, -1, embed_dim
        )

        return final_embeddings
```

Reply:
@mzbac are you going to add the snippet you shared?

Reply:
@awni, I don't have permission to push the changes. Maybe either you or @nkasmanoff could update it.

Reply:
😬 sorry. Yea happy to, or @nkasmanoff can give you temporary permissions to push to his fork.

Reply:
@mzbac just gave you permission, let me know if that worked!

Reply:
@nkasmanoff @awni I have pushed the changes. Please note that the test.py will fail due to the fast_gelu to native gelu change, but if you build the latest mlx locally, the test will pass.

Reply:
Hmm, we should put the version in the requirement in that case. Should work with `0.5` and up.